[receiver/googlecloudmonitoringreceiver] Transform GCP Timeseries Data to OTel Metrics Conversion Pipeline

abhishek-at-cloudwerx opened the pull request:
**Description:** Transform GCP Timeseries Data to OTel Metrics Pipeline by establishing a client connection to the Google Cloud Monitoring API. This PR scrapes timeseries data based on configured metrics and converts it into the OpenTelemetry format for pipeline processing.

**Link to tracking Issue:** #33762

**Testing:** No additional test cases were added. Existing tests cover the necessary functionality.

**Documentation:** The **README.md** file was previously added and already includes relevant details about the component and the conversion process.

abhishek-at-cloudwerx committed:
Feat:
- Create client connection for monitoring REST API
- Implement logic for fetch metrics data and convert it into pipeline data

abhishek-at-cloudwerx committed:
Feat:
- Auto import

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Fix:
- Fix fail check issues

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
-	if err != nil {
-		return err
-	}
+func (mr *monitoringReceiver) Shutdown(context.Context) error {
```
Shutdown should probably Close() the metric client.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+
+		// Log an error if the filter query is empty
+		if filterQuery == "" {
+			mr.logger.Error("Internal Server Error")
```
we should probably `continue` here so that we don't make the query

abhishek-at-cloudwerx replied:
Remove the condition, as validation on the config variable will directly trigger an error from the test cases.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+			timeSeriesMetrics, err := it.Next()
+			if timeSeriesMetrics == nil && err != nil {
+				if errors.Is(err, iterator.Done) {
+					mr.logger.Info(iterator.Done.Error())
```
Should we only log this error if it is not nil?  The example seems to assume the iterator.Done error indicates we successfully finished iterating: https://pkg.go.dev/cloud.google.com/go/monitoring/apiv3/v2#MetricClient.ListTimeSeries

dashpole replied:
I think you might be able to simplify this block to just 
```golang
if errors.Is(err, iterator.Done) {
    break
}
```

abhishek-at-cloudwerx replied:
Agree

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+
+		// Create an iterator for the time series data
+		it := mr.client.ListTimeSeries(ctx, req)
+		mr.logger.Info("Time series data:")
```
you don't actually log the data here?  This should probably also be at debug level.

abhishek-at-cloudwerx replied:
That's correct; it was just an informational message and has now been set to the debug level.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+			// Handle errors and break conditions for the iterator
+			if err != nil {
+				err := fmt.Errorf("failed to retrieve time series data: %w", err)
+				gErr = multierr.Append(gErr, err)
```
The Go stdlib now has support for multiple errors.  Consider using errors.Join instead of multierr.

dashpole replied:
Multi error joining is also unneccessary here.  You return right away, so there is only ever a single error returned.

abhishek-at-cloudwerx replied:
Updated

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+	now := time.Now()
+
+	// Calculate the start time (current time - delay)
+	startTime := now.Add(-delay - interval)
```
nit: would be simpler to do `endTime := now.Add(-delay)` and `startTime := endTime.Add(-interval)`?

abhishek-at-cloudwerx replied:
Yes True.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+		m.SetUnit(resp.Unit)
+
+		// Assuming MetricDescriptor and description are set
+		m.SetDescription("Converted from GCP Monitoring TimeSeries")
```
You will probably need to list Metric Descriptors to get descriptions.  But you can cache the response, at least.  Please add a TODO.

abhishek-at-cloudwerx replied:
I added a TODO, but when I tried to get the Metric Description, I received 'google.api.Metric' by fetching resp.Metric.ProtoReflect().Descriptor().FullName().

abhishek-at-cloudwerx replied:
So, this is similar to what we did when listing the time series data, correct?

dashpole replied:
Yes.  The difference is that ListMetricDescriptors is not designed to be called repeatedly.  You should only call it once, and cache the results.

abhishek-at-cloudwerx replied:
We can't fetch the entire metrics data at once; otherwise, we'd have to differentiate between them using if-else conditions. To avoid that, we're using filters to fetch specific metric descriptors' data. However, with filters, we only retrieve data for a single metric descriptor at a time, so the same iteration will occur. Is it okay to handle it similarly to how we process time series data?

dashpole replied:
That should be good to start

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+		m.SetDescription("Converted from GCP Monitoring TimeSeries")
+
+		// Set resource labels
+		resource := rm.Resource()
```
Won't this end up clobbering the resource if multiple different timeseries are returned?  Each loop we are editing the same resource.  Some potential approaches to fix this:

* Sort timeSeriesMetrics by monitored resource, and then iterate over the grouped timeSeriesMetrics
* Keep a lookup table for GCP monitored resource -> OTel resource so we only create a new resource if it is the first time we've seen that monitored resource.

abhishek-at-cloudwerx replied:
The issue is resolved because now we handle each metric individually rather than looping through a bunch of time-series metrics.

dashpole replied:
We still appear to be looping through time-series metrics.  Did you forget to push changes?

abhishek-at-cloudwerx replied:
Now updated, please have a look.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+			mr.metricsBuilder.ConvertSumToMetrics(resp, m)
+		case metric.MetricDescriptor_DELTA:
+			mr.metricsBuilder.ConvertDeltaToMetrics(resp, m)
+		// Add cases for SUMMARY, HISTOGRAM, EXPONENTIAL_HISTOGRAM if needed
```
GCP doesn't have a summary type, so no need to worry about that.  You can add TODOs for histogram and exponential histogram.

abhishek-at-cloudwerx replied:
Added a TODO.

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go:
```diff
@@ -0,0 +1,80 @@
+
+	for _, point := range ts.GetPoints() {
+		dp := gauge.DataPoints().AppendEmpty()
+		dp.SetStartTimestamp(pcommon.Timestamp(point.Interval.StartTime.Seconds))
```
Yes, that's correct. We can add a conditional statement: if the value exists, we set it; otherwise, we leave it unset.

Do we need to do the same thing for endtime as well ?

dashpole replied:
Yes. I would do it just for safety.  It shouldn't ever happen, so you can log a warning and skip the point if no end time is set.

abhishek-at-cloudwerx committed:
Refactor:
- Add cases for metrics value in SUM and DELTA metrics conversion functions

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go:
```diff
@@ -0,0 +1,80 @@
+		dp := sum.DataPoints().AppendEmpty()
+		dp.SetStartTimestamp(pcommon.Timestamp(point.Interval.StartTime.Seconds))
+		dp.SetTimestamp(pcommon.Timestamp(point.Interval.EndTime.Seconds * 1e9)) // Convert to nanoseconds)
+		dp.SetDoubleValue(point.GetValue().GetDoubleValue())
```
Updated with cases

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go:
```diff
@@ -0,0 +1,80 @@
+		dp := sum.DataPoints().AppendEmpty()
+		dp.SetStartTimestamp(pcommon.Timestamp(point.Interval.StartTime.Seconds))
+		dp.SetTimestamp(pcommon.Timestamp(point.Interval.EndTime.Seconds * 1e9)) // Convert to nanoseconds
+		dp.SetDoubleValue(point.GetValue().GetDoubleValue())
```
Updated with cases

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go:
```diff
@@ -0,0 +1,80 @@
+
+	for _, point := range ts.GetPoints() {
+		dp := gauge.DataPoints().AppendEmpty()
+		dp.SetStartTimestamp(pcommon.Timestamp(point.Interval.StartTime.Seconds))
```
Updated

dashpole replied:
Please do the same thing for the timestamp as well

abhishek-at-cloudwerx replied:
Done

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Refactor:
- Refactor code based on PR comments

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,203 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
-	m.logger.Debug("Scrape metrics ")
+func (mr *monitoringReceiver) Start(ctx context.Context, _ component.Host) error {
+	var startErr error
+	mr.startOnce.Do(func() {
```
Yes, I implemented a retry function with a 5-second delay.

dashpole replied:
You should leave retry logic to the underlying collector framework.  I just meant to say that a sync.Once won't work here because we may need this to run again if it failed the first time.

abhishek-at-cloudwerx replied:
Got it, Updated Now

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/testdata/config.yaml:
```diff
@@ -1,8 +1,8 @@
   metrics_list:
     - metric_name: "compute.googleapis.com/instance/cpu/usage_time"
-      delay: 60s # Second
+      delay: 1m # Can be specified in seconds (s), minutes (m), or hours (h)
```
I have decided to use the term "fetch_delay" for this parameter, as it reflects the delay in ingesting older data and set its default value to 1 minute.

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Fix:
- Fix test-case issue by adding service.json file on testdata folder
- Setup env variable for GCP service file
- Remove retry logic and sync.Once for client connection

abhishek-at-cloudwerx committed:
Fix:
- Fix lint issue for googlecloudmonitoring receiver component

dashpole commented on receiver/googlecloudmonitoringreceiver/go.mod:
```diff
@@ -4,22 +4,41 @@ go 1.22.6
-	go.opentelemetry.io/collector/consumer/consumertest v0.107.1-0.20240816132030-9fd84668bb02
-	go.opentelemetry.io/collector/pdata v1.13.1-0.20240816132030-9fd84668bb02
-	go.opentelemetry.io/collector/receiver v0.107.1-0.20240816132030-9fd84668bb02
+	go.opentelemetry.io/collector/component v0.107.0
```
you will need to revert these changes, which downgrade the collector version.

abhishek-at-cloudwerx replied:
Yes, updated

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Fix:
- Fix version issue in go.sum and error type

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,210 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+	}
+
+	// Convert the GCP TimeSeries to pmetric.Metrics format of OpenTelemetry
+	metrics := mr.convertGCPTimeSeriesToMetrics(allTimeSeriesMetrics)
```
Yes, exactly. However, the convertGCPTimeSeriesToMetrics function returns a pmetric.Metric object, and we can't return a response for just a single metric. In the end, we'll need to append all the converted metric responses together before returning them to the exporter, so it will ultimately consume the same amount of memory.

dashpole replied:
you could pass the pmetric.Metric as an argument to the convertGCPTimeSeriesToMetrics, and have it append to that if you want. You will eventually need to store the resulting `pmetric.Metric`, but you don't need to keep a reference to all of the responses at once.  If you aren't convinced we can omit it from this PR

abhishek-at-cloudwerx replied:
Understood and updated now.

abhishek-at-cloudwerx committed:
Fix:
- Changes based on PR comments for mutex and timestamp

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,210 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
-	return metrics, nil
-}
+	// Lock to ensure thread-safe access to mr.client
+	mr.mutex.Lock()
```
Updated

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

dashpole commented on receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go:
```diff
@@ -0,0 +1,131 @@
+		case *monitoringpb.TypedValue_Int64Value:
+			dp.SetIntValue(v.Int64Value)
+		default:
+			log.Printf("Unhandled metric value type: %T", v)
```
please use the logger instead of the log package.  Same below for other types.

abhishek-at-cloudwerx replied:
Updated

dashpole commented on receiver/googlecloudmonitoringreceiver/internal/metrics_conversion.go:
```diff
@@ -0,0 +1,131 @@
+		} else {
+			warnMsg := fmt.Sprintf("EndTime is invalid for metric: %s", ts.GetMetric().GetType())
+			mb.logger.Warn(warnMsg)
+			continue
```
Actually, should we delete the datapoint if it is invalid?  Otherwise we will be sending bad data here.  The same question applies to an invalid value type below

abhishek-at-cloudwerx replied:
I believe there's no need to delete the data point values since we're already skipping them, which ensures that no incorrect data will be sent further.

dashpole replied:
The way the pmetric API works is that once you call AppendEmpty(), it has already been added to the batch.  Then, you just fill in details.  By continuing here, you are just leaving the metric as-is.

abhishek-at-cloudwerx replied:
Updated

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,229 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+	}
+
+	// Get service account key file path
+	serAccKey := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
```
You should use https://pkg.go.dev/golang.org/x/oauth2/google#FindDefaultCredentials to get credentials, rather than looking up the environment variables yourself.

abhishek-at-cloudwerx replied:
Yes Updated Now

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,229 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
-	return nil
+func (mr *monitoringReceiver) Shutdown(context.Context) error {
+	var err error
+	if mr.client != nil {
```
I think you should probably lock around this.

abhishek-at-cloudwerx replied:
Updated

abhishek-at-cloudwerx committed:
Feat:
- Add function to fetch a list of metric descriptors
- Call function to convert time series metrics data once

abhishek-at-cloudwerx committed:
Fix:
- Apply zap logger instead of default log package

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,268 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+
+// ConvertGCPTimeSeriesToMetrics converts GCP Monitoring TimeSeries to pmetric.Metrics
+func (mr *monitoringReceiver) convertGCPTimeSeriesToMetrics(metrics pmetric.Metrics, metricDesc *metric.MetricDescriptor, timeSeries *monitoringpb.TimeSeries) {
+	rm := metrics.ResourceMetrics().AppendEmpty()
```
This is more correct than it was previously, but still not optimal.  It will now create a new ResourceMetrics for each TimeSeries.  This can result in duplicate resources within a batch of ResourceMetrics if multiple TimeSeries are associated with the same MonitoredResource (i.e. have the same UserLabels and SystemLabels).  Ideally, if this is a resource we have seen before, we would append a new Metric to the existing ResourceMetric.

E.g.
TimeSeries:
  - Resource=foo, Metric=bar
  - Resource=foo, Metric=baz
  - Resource=foo, Metric=dong
  
  currently becomes:
  
resourcemetrics:
- resource: foo
   scopemetrics:
   metrics:
   - name: bar
- resource: foo
   scopemetrics:
   metrics:
   - name: baz
- resource: foo
   scopemetrics:
   metrics:
   - name: dong

We would instead prefer:

resourcemetrics:
- resource: foo
   scopemetrics:
   metrics:
   - name: bar
   - name: baz
   - name: dong

abhishek-at-cloudwerx committed:
Fix:
- Use Default credentials inbuilt function for finding a service.json file
- Make metrics descriptors variable to common
- Arrange resource metrics response

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,268 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudmonitoringreceiver/internal"
 )
 
+var allMetricDescriptor = make(map[string]*metric.MetricDescriptor)
```
Updated

abhishek-at-cloudwerx commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,268 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+
+	// Set resource labels
+	resource := rm.Resource()
+	resource.Attributes().PutStr("resource_type", timeSeries.Resource.Type)
```
Updated

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,316 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+	var err error
+
+	// Use google.FindDefaultCredentials to find the credentials
+	creds, _ := google.FindDefaultCredentials(ctx)
```
You should probably return the error from this function, rather than ignore it.

abhishek-at-cloudwerx replied:
Actually it will fail the auto generated test cases that's why i am not returning it

dashpole replied:
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/34677#discussion_r1743871226.  We should disable those tests, and properly handle the error here.

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,316 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+		client, err = monitoring.NewMetricClient(ctx, option.WithCredentials(creds))
+	} else {
+		// Set a default credentials file path for testing
+		os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "testdata/serviceAccount.json")
```
You should do this in test code, rather than in the main logic.

dashpole replied:
You should set this in metadata.yaml to skip tests that require auth setup:

```
tests:
  skip_lifecycle: true
```

dashpole replied:
Then remove this from the receiver.

abhishek-at-cloudwerx replied:
Removed

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,316 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+
+	// If the client is already initialized, return nil
+	if mr.client != nil {
+		return nil
```
If we successfully created the client last time, but failed to get metric descriptors, we should try to get metric descriptors again.

abhishek-at-cloudwerx replied:
In this case we can add a retry mechanism for getting metric descriptors. 
For exa- Max Retries = 3 and It will call after every 5 seconds.

abhishek-at-cloudwerx replied:
Should I implement a retry mechanism?

dashpole replied:
No.  But you shouldn't return here if metric descriptors is empty.  You should skip client initialization if the client is non-nil, but then also check to see if metric descriptors are populated before skipping that part.

You could make initializeClient() and initializeMetricDescriptors() separate functions if you wanted.

abhishek-at-cloudwerx replied:
Updated with separate functions

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Fix:
- Make changes

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Fix:
- Fix go.sum file changes

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Merge branch 'main' into googlecloudmonitoringreceiver-phase2

abhishek-at-cloudwerx committed:
Refactor:
- Skip authentication setup test cases
- Separate functions for initializing the monitoring client and descriptor API

dashpole commented:
@TylerHelmuth this lgtm if you want to take a look

TylerHelmuth commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,328 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+			if err != nil {
+				return fmt.Errorf("failed to retrieve metric descriptors data: %w", err)
+			}
+			mr.metricDescriptors[metricDesc.Type] = metricDesc
```
As far as I can tell, there is a write here and a read on line 94 that is not protected by a lock.  It would be strange for Start and Scrape to be called simultaneously, but do we need to protect it anyways?

abhishek-at-cloudwerx replied:
Resolved now with locking it

abhishek-at-cloudwerx committed:
Fix:
- Resolved issue with metric descriptor retrieval
- Fixed monitoring client connection problem

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,331 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+			gInternal = defaultCollectionInterval
+		}
+
+		gDelay = metric.FetchDelay
```
I learned from another Googler that the fetch delay is actually provided by the metric descriptor: https://pkg.go.dev/google.golang.org/genproto/googleapis/api/metric#MetricDescriptor_MetricDescriptorMetadata

You can get it with
```golang
gDelay = metricDesc.GetMetadata().GetIngestDelay().AsDuration()
```

You can actually remove this from the configuration as well.

abhishek-at-cloudwerx replied:
Remove the **fetch_delay** from the configuration file and instead retrieve it from the metric descriptor.

dashpole commented on receiver/googlecloudmonitoringreceiver/config.go:
```diff
@@ -11,7 +11,10 @@ import (
 
-const minCollectionIntervalSeconds = 60
+const (
+	defaultCollectionInterval = 60 * time.Second // Default value for collection interval
```
According to another Googler who works on the monitoring API, we should use a defualt value of 5 minutes for the collection interval.

abhishek-at-cloudwerx replied:
Updated

dashpole commented:
One other piece of feedback from the Google cloud monitoring team:  We should make sure we never query for data more than 24h in the past, as that can cause performance issues for the API.  We should either validate that the interval configuration option is less than ~23 hours, or just enforce a max of ~23 hours when we calculate the interval.

abhishek-at-cloudwerx committed:
Refactor:
- Remove fetch_delay from config file
- Set default collection interval value and max value of 23 hours

dashpole commented on receiver/googlecloudmonitoringreceiver/receiver.go:
```diff
@@ -5,51 +5,339 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
+}
+
+func (mr *monitoringReceiver) Start(ctx context.Context, _ component.Host) error {
+	mr.metricDescriptors = make(map[string]*metric.MetricDescriptor)
```
lets do this in newGoogleCloudMonitoringReceiver above, rather than here so we don't overwrite mr.metricDescriptors if start is called again?

abhishek-at-cloudwerx replied:
Moved now

abhishek-at-cloudwerx committed:
Refactor:
- Changes based on PR

abhishek-at-cloudwerx committed:
Fix:
- Fix config test cases
